src/axolotl/state.rs

```diff
@@ -122,24 +97,49 @@ impl <T:Axolotl> ReceiveChain<T> {
 
         for i in self.chain_key_index..(index+1) {
             let (next_chain_key, message_key) = T::derive_next_chain_and_message_key(&self.chain_key);
-            self.chain_key = next_chain_key;
             self.message_keys.push((i,message_key));
+            self.chain_key = next_chain_key;
+            self.chain_key_index += 1;
         }
-        self.chain_key_index = index+1;
 
-        return Some(self.message_keys.pop().unwrap().1);
+        return Some(self.message_keys.len()-1);
+    }
+
+    fn try_decrypt(
+        &mut self,
+        message : &AxolotlMessage<T>, 
+        mac : T::Mac,
+        sender_identity : &DHPublic<T::IdentityKey>, 
+        receiver_identity : &DHPublic<T::IdentityKey>,
+    ) -> Option<T::PlainText> {
+        self.try_get_message_key_index(message.message_number)
+            .and_then(|message_key_index| {
+                let (_,ref message_key) = self.message_keys[message_key_index];
+                let expected_mac = T::authenticate_message(message, message_key, sender_identity, receiver_identity);
+                if expected_mac == mac {
+                    T::decrypt_message(&message_key, &message.ciphertext)
+                        .map(|plaintext| (message_key_index, plaintext))
+                }
+                else {
+                    None
+                }
+            })
+            .map(|(message_key_index,plaintext)| {
```

Comment:
This function could probably be cleaned up. Took me a while to follow what it was doing. The Option chaining is a nice, but decomposing into functions would make it more readable